### PR TITLE
Migrate doc building to github action workflows

### LINKED ISSUE #4367: Migrate doc building to github action workflows
Currently docs are built with gdal/doc/.azure-pipelines.yml. It would be good to consolidate on using github workflow instead (inspiration can be taken from https://github.com/OSGeo/PROJ/blob/3bb6c1bf02c454686edf557a520ae6c70d6ae5d9/.github/workflows/doc_build.yml#L34), and for example use the possibility to upload artifacts so we can download the generated PDF / HTML (useful for pull requests that modify the doc). See https://github.com/OSGeo/PROJ/blob/3bb6c1bf02c454686edf557a520ae6c70d6ae5d9/.github/workflows/doc_build.yml#L83 for that last point.

### PR DESCRIPTION
**Pull request by cmarmo**
## What does this PR do?
This pull request migrates the documentation building process to github actions.
Thanks for considering it.

## What are related issues/pull requests?
It fixes #4367.

## Notes
Should the azure pipeline be removed in this pull request too?
The 'spelling' step implemented in https://github.com/OSGeo/PROJ/blob/c80c4ededbad6a11e9731ae5e118f7cfd9138f67/.github/workflows/doc_build.yml#L92 fails due to some spelling warnings. I have commented it out for now. It wasn´t implemented in azure. Should it be removed?

### CODE REVIEW

**Comment by rouault** [comment]
Thanks for having a look at that. 

> Should the azure pipeline be removed in this pull request too?

That would be the ultimate goal, but there's one key thing from the Azure pipeline that this PR doesn't cover: for builds of the master branch of OSGeo/gdal, the upload to https://github.com/OSGeo/gdal-docs which is what powers gdal.org . We'd likely need to use github secrets (https://docs.github.com/en/actions/security-guides/encrypted-secrets) to convey a SSH private key to git push to that repository. (an example of using secrets is in https://github.com/OSGeo/PROJ/blob/master/.github/workflows/docker.yml to push to a docker repository)
For that, you'd have to prototype that first in your own fork (using github actions on push in your fork, not in this PR), adding in its settings a secret with a private key of yours and push in a dummy gdal-docs repository in your fork, and once that's working, we could transpose that here and set a secrets in OSGeo/gdal with a SSH private key allowed to push to OSGeo/gdal-docs. Hope that makes sense...

**Comment by cmarmo** [comment]
> > Should the azure pipeline be removed in this pull request too?
> 
> That would be the ultimate goal, but there's one key thing from the Azure pipeline that this PR doesn't cover: for builds of the master branch of OSGeo/gdal, the upload to https://github.com/OSGeo/gdal-docs which is what powers gdal.org .

Thanks for your answer and your clarifications.
Yes, I've seen the upload step to gdal.org in azure, I was not sure I was allowed to propose PRs touching the 'secret' gdal infrastructure... :)
I will manage to simulate the process on my side and update the PR then.

**Comment by cmarmo** [comment]
Thanks for your patience.
I have tested the current version and it worked on my side. I have modified user and email git configuration as GDAL bot as you might want to reuse the keys on azure.
In this configuration the private ssh key should be added as GDAL repository secret (SSH_KEY_DOCS in my config), the public key should be added as deploy key in the gdal-docs repository.
Let me know if this is what you had in mind. Thanks.

File: .github/workflows/doc_build.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    - uses: actions/upload-artifact@v2
+      with:
+        name: HTML
+        path: doc/build/html/*
+    #- uses: actions/upload-artifact@v2
+    #  with:
+    #    name: Misspelled
+    #    path: doc/build/spelling/output.txt
+    - name: Deploy ssh key
+      if: ${{ github.ref_name == 'master' }}
```
Comment (rouault) [inline .github/workflows/doc_build.yml:64]:
```suggestion
      if: ${{ github.ref_name == 'master' && github.repository == 'OSGeo/gdal' }} 
```

File: .github/workflows/doc_build.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      shell: bash -l {0}
+      run: |
+        mkdir /root/.ssh && echo "${{ secrets.SSH_KEY_DOCS }}" > /root/.ssh/id_rsa
+        wc /root/.ssh/id_rsa
+        chmod 700 /root/.ssh && chmod 600 /root/.ssh/id_rsa
+        ssh-keyscan -t rsa github.com >> /root/.ssh/known_hosts
+        eval `ssh-agent -s`
+        ssh-add /root/.ssh/id_rsa
+    - name: Deploy to gdal.org
+      if: ${{ github.ref_name == 'master' }}
```
Comment (rouault) [inline .github/workflows/doc_build.yml:74]:
```suggestion
      if: ${{ github.ref_name == 'master' && github.repository == 'OSGeo/gdal' }}
```

**Comment by rouault** [comment]
@hobu Would you mind adding a SSH_KEY_DOCS secrets with the content of the ssh private key that corresponds to the $DOWNLOADSECUREFILE_SECUREFILEPATH used by doc/.azure-pipelines.yml ?

File: .github/workflows/doc_build.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    #- uses: actions/upload-artifact@v2
+    #  with:
+    #    name: Misspelled
+    #    path: doc/build/spelling/output.txt
+    - name: Deploy ssh key
+      if: ${{ github.ref_name == 'master' }}
+      shell: bash -l {0}
+      run: |
+        mkdir /root/.ssh && echo "${{ secrets.SSH_KEY_DOCS }}" > /root/.ssh/id_rsa
+        wc /root/.ssh/id_rsa
```
Comment (cmarmo) [inline .github/workflows/doc_build.yml:68]:
```suggestion
```
I'm removing this line... this was for debugging... sorry

**Comment by cmarmo** [comment]
For the sake of completeness [here](https://github.com/cmarmo/gdal/runs/4988123492?check_suite_focus=true) is the log of my tests.

**Comment by hobu** [comment]
`SSH_KEY_DOCS` secret added to repository. Note that repository secrets are not available to pull requests.

**Comment by rouault** [comment]
@hobu The job after the merge fails: https://github.com/OSGeo/gdal/runs/5023341722?check_suite_focus=true : " ERROR: Permission to OSGeo/gdal-docs.git denied to proj4docs." .

**Comment by cmarmo** [comment]
> the public key should be added as deploy key in the gdal-docs repository.

May I ask if the public key has been deployed?

**Comment by rouault** [comment]
>  May I ask if the public key has been deployed?

no, https://github.com/OSGeo/gdal-docs/settings/keys is empty for now . The repository has just a single user with write rights , https://github.com/pdal-docs , and I suspect the SSH_KEY_DOCS secret must be its ssh key.
Why would a deploy key be necessary compared to the Azure pipeline job ? I see you've added a "ssh-add /root/.ssh/id_rsa" step that wasn't in the Azure pipeline. Would that be related ?

**Comment by cmarmo** [comment]
> > May I ask if the public key has been deployed?
> 
> no, https://github.com/OSGeo/gdal-docs/settings/keys is empty for now . The repository has just a single user with write rights , https://github.com/pdal-docs , and I suspect the SSH_KEY_DOCS secret must be its ssh key. Why would a deploy key be necessary compared to the Azure pipeline job ? I see you've added a "ssh-add /root/.ssh/id_rsa" step that wasn't in the Azure pipeline. Would that be related ?

I have write rights on my simulated doc repository, but installing the public key in my ssh-keys was not enough to push.... Unfortunately I only can make some reverse engineering on the github infrastructure. This works if the public key is installed as deploy key in the doc repo.
I added the `ssh-add` because otherwise I was asked for a passphrase.... Again, I'm reverse engineering.

**Comment by cmarmo** [comment]
Please note that I have debugged the entire workflow with
```
ssh -T git@hostname
```
before trying to push, as documented [here](https://docs.github.com/en/github-ae@latest/authentication/connecting-to-github-with-ssh/testing-your-ssh-connection)

**Comment by rouault** [comment]
ok, I'm going to try with a freshly new generated key and will update SSK_KEY_DOCS and add the public key as a deploy key to gdal-docs

**Comment by rouault** [comment]
ok, that did it: https://github.com/OSGeo/gdal-docs has now been refreshed ! . thanks @cmarmo

**Comment by cmarmo** [comment]
Me again.... is that normal that the page deployment failed?
https://github.com/OSGeo/gdal-docs/deployments/activity_log?environment=github-pages

**Comment by rouault** [comment]
The issue on the build stage of the deployment https://github.com/OSGeo/gdal-docs/runs/5027754499?check_suite_focus=true : "tar: ./gdal.pdf: File removed before we read it".  I suspect this might be due to an empty file. Comparing again the new workflow with the Azure pipelines, I noticed there were missing steps. I've re-added them per f9b972f0e4 . untested for now. let's see what the next run will tell

**Comment by cmarmo** [comment]
Ok, sorry, I didn't realize that other things than the build/html directory were uploaded.

**Comment by rouault** [comment]
> let's see what the next run will tell

that's fixed it. latest deployment was successful: https://github.com/OSGeo/gdal-docs/actions

**Pull request merged**